src/string_pool.cr

```diff
@@ -119,11 +151,22 @@ class StringPool
       next_probe_offset += 1_u64
     end
 
-    @size += 1
-    entry = String.new(str, len)
-    @hashes[index] = hash
-    @values[index] = entry
-    entry
+    yield index
+  end
+
+  private def get(hash : UInt64, str : UInt8*, len)
+    get(hash, str, len) do |index|
+      rehash if @size >= @capacity // 4 * 3
+      @size += 1
+      entry = String.new(str, len)
+      @hashes[index] = hash
+      @values[index] = entry
+      entry
+    end
+  end
+
+  private def get?(hash : UInt64, str : UInt8*, len)
+    get(hash, str, len) { nil }
   end
```

Comment:
**polish**: the method is private and called exactly once, it could be inlined right in `get?(ptr, len)`.